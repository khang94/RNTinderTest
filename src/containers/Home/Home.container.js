import React, {Component} from 'react';
import {View, Text} from 'react-native';
import PropTypes from 'prop-types';
import DeckSwiper from 'react-native-deck-swiper';

import {connect} from 'react-redux';

import {actions} from '../../redux/Favourite/Favourite.reducer';
import styles from './Home.styles';
import userJson from './../../assets/json/sample.json';
import Card from '../../components/Card/Card.component';
import CardWithLoading from '../../components/Card/CardWithLoading.component';
import Handlers from './Home.handlers';
import Constants from '../../config/Constants';

const {Direction} = Constants;

class HomeContainer extends Component {
  constructor(props) {
    super(props);
    this.handleState = this.setState.bind(this);
  }

  state = {
    users: [],
    currentUser: userJson.results[0].user,
    activeTab: 'location',
    loading: false,
  };

  componentDidMount() {
    Handlers.fetchUser(this.handleState, this.state);
  }

  renderCard = card => {
    const {currentUser, loading} = this.state;
    const avatar = (currentUser && currentUser.picture) || '';
    return loading ? (
      <CardWithLoading />
    ) : (
      <Card avatar={avatar} user={currentUser} />
    );
  };

  onSwipedRight = () => {
    this.saveFavouriteUser();
  };

  onSwipedLeft = () => {
    Handlers.fetchUser(this.handleState, this.state);
  };

  saveFavouriteUser = () => {
    const {currentUser} = this.state;
    const {addFavouriteUser} = this.props;
    addFavouriteUser({user: currentUser});
  };

  render() {
    const {users} = this.state;
    return (
      <View style={styles.container}>
        <View style={styles.top} />
        <Text style={styles.title}>Find Your Friends</Text>

        <DeckSwiper
          containerStyle={styles.card}
          cards={users}
          renderCard={this.renderCard}
          onSwipedLeft={this.onSwipedLeft}
          onSwipedRight={this.onSwipedRight}
          cardIndex={0}
          backgroundColor={'transparent'}
          stackSize={1}
          keyExtractor={(user, index) => user + index}
        />
      </View>
    );
  }
}

HomeContainer.defaultProps = {
  addFavouriteUser: () => {},
};

HomeContainer.propTypes = {
  addFavouriteUser: PropTypes.func,
};

const mapDispatchToProps = {
  addFavouriteUser: actions.addFavouriteUser,
};

export {HomeContainer};

export default connect(
  null,
  mapDispatchToProps,
)(HomeContainer);
