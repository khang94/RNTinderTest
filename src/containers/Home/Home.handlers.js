import {getUserAPI} from './../../services/fetchUserApi';

const showLoading = handleState => {
  handleState({
    loading: true,
  });
};

const fetchUser = (handleState, state) => {
  showLoading(handleState);
  getUserAPI().then(res => {
    if (res && res.results[0]) {
      const {user} = res.results[0];

      const updatedUsers = [...state.users, user];
      handleState({
        users: updatedUsers,
        currentUser: user,
        loading: false,
      });
    }
  });
};

export default {
  showLoading,
  fetchUser,
};
