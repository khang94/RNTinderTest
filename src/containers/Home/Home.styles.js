import {StyleSheet} from 'react-native';
import {Dimensions} from 'react-native';

const {width, height} = Dimensions.get('screen');

export default StyleSheet.create({
  container: {
    backgroundColor: '#BDBDBD',
    height: height,
  },
  top: {
    height: height / 3,
    width,
    backgroundColor: 'grey',
  },
  title: {
    top: 30,
    fontSize: 20,
    fontWeight: 'bold',
    position: 'absolute',
    alignSelf: 'center',
  },
  card: {
    marginTop: 100,
    height: 350,
    marginHorizontal: 10,
    borderRadius: 5,
  },
  topCard: {
    paddingTop: 10,
    borderTopLeftRadius: 5,
    borderTopRightRadius: 5,
    backgroundColor: '#E0E0E0',
    height: 100,
    borderBottomColor: '#9E9E9E',
    borderBottomWidth: 1,
    alignItems: 'center',
  },
});
