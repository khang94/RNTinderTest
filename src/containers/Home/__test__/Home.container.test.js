import React from 'react';

import {HomeContainer} from '../Home.container';
import {shallow} from 'enzyme';

import Handlers from '../Home.handlers';

describe('HomeContainer snapshot', () => {
  const props = {
    addFavouriteUser: jest.fn(),
  };

  const user = {
    name: {
      first: 'Steve',
      last: 'Job',
    },
  };

  it('Should render snapshot', () => {
    const wrapper = shallow(<HomeContainer {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should save to favourite user store if swipe right', () => {
    const wrapper = shallow(<HomeContainer {...props} />);

    wrapper.setState({currentUser: user});
    wrapper.instance().onSwipedRight();
    expect(props.addFavouriteUser).toBeCalledWith({user});
  });

  it('Should save to favourite user store if swipe left', () => {
    const wrapper = shallow(<HomeContainer {...props} />);

    Handlers.fetchUser = jest.fn();
    wrapper.setState({currentUser: user});
    wrapper.instance().onSwipedLeft();
    expect(Handlers.fetchUser).toBeCalled();
  });

  it('Should render card item based on current user data after fetched ', () => {
    const wrapper = shallow(<HomeContainer {...props} />);
    wrapper.setState({currentUser: user, loading: false});

    expect(wrapper.instance().renderCard()).toMatchSnapshot();
  });

  it('Should render loading state in card while fetching data ', () => {
    const wrapper = shallow(<HomeContainer {...props} />);
    wrapper.setState({currentUser: user, loading: true});

    expect(wrapper.instance().renderCard()).toMatchSnapshot();
  });
});
