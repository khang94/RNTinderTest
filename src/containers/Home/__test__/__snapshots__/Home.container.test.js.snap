// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeContainer snapshot Should render card item based on current user data after fetched  1`] = `
<Card
  avatar=""
  user={
    Object {
      "name": Object {
        "first": "Steve",
        "last": "Job",
      },
    }
  }
/>
`;

exports[`HomeContainer snapshot Should render loading state in card while fetching data  1`] = `<CardWithLoading />`;

exports[`HomeContainer snapshot Should render snapshot 1`] = `ShallowWrapper {}`;
